File: scripts/detectHosViolations.mjs
#!/usr/bin/env node
/**
 * HOS / ELD Compliance Watchdog — queries bcat-ops for driver compliance alerts,
 * tracks 7-day rolling violation counts, and posts violations to Slack.
 *
 * READ-ONLY. Queries the bcat-ops AppSync GraphQL API for:
 *   - ComplianceAlert records (unacknowledged, CRITICAL/EXPIRED)
 *   - Driver records with NON_COMPLIANT/EXPIRING_SOON complianceStatus
 * Tracks daily violation counts per driver in a 7-day rolling state file.
 *
 * Auth: Cognito signIn via aws-amplify (same auth as detectFuelAnomalies.mjs).
 *
 * Usage:
 *   source ~/.hermes/bcat-ops-credentials.env
 *   node scripts/detectHosViolations.mjs [--json] [--slack] [--quiet]
 *
 * State file: ~/.hermes/state/hos-watchdog-state.json
 *
 * For cron: wrapper at ~/.hermes/scripts/hos-watchdog.sh
 */

import { readFileSync } from 'fs'
import { dirname, resolve } from 'path'
import { homedir } from 'os'
import { fileURLToPath } from 'url'
import { Amplify } from 'aws-amplify'
import { signIn, fetchAuthSession } from 'aws-amplify/auth'

const __dirname = dirname(fileURLToPath(import.meta.url))
const HOME = homedir()

// ── Amplify config ─────────────────────────────────────────────────────────────
const outputs = JSON.parse(readFileSync(resolve(__dirname, '../amplify_outputs.json'), 'utf8'))
const APPSYNC_URL = outputs.data.url
Amplify.configure(outputs)

// State file for 7-day rolling violation tracking
const STATE_FILE = `${HOME}/.hermes/state/hos-watchdog-state.json`

// ── GraphQL ────────────────────────────────────────────────────────────────────
const LIST_ALERTS_QUERY = `query ListComplianceAlerts {
  listComplianceAlerts(limit: 1000) {
    items {
      id entityType entityId entityName documentType documentTitle
      complianceDocumentId expirationDate severity acknowledged
      acknowledgedBy acknowledgedAt resolvedAt createdAt updatedAt
    }
  }
}`

const LIST_DRIVERS_QUERY = `query ListDrivers {
  listDrivers(limit: 200) {
    items {
      id name active type onboardingStatus complianceStatus
      cdl cdlExpiration medCardExpiration hireDate
      assignedTruckId email phone createdAt updatedAt
    }
  }
}`

async function callAppSync(query, variables, idToken) {
  const res = await fetch(APPSYNC_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: idToken },
    body: JSON.stringify({ query, variables }),
  })
  return res.json()
}

// ── State file management ──────────────────────────────────────────────────────

import { writeFileSync, existsSync, mkdirSync } from 'fs'

function loadState() {
  if (existsSync(STATE_FILE)) {
    try {
      return JSON.parse(readFileSync(STATE_FILE, 'utf8'))
    } catch {
      // corrupt file, start fresh
    }
  }
  return { daily_counts: {}, last_updated: null }
}

function saveState(state) {
  const dir = resolve(STATE_FILE, '..')
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true })
  state.last_updated = new Date().toISOString()
  writeFileSync(STATE_FILE, JSON.stringify(state, null, 2))
}

function updateRollingCounts(state, todayIso, violationDrivers) {
  const daily = state.daily_counts || {}
  daily[todayIso] = {
    drivers: [...violationDrivers].sort(),
    count: violationDrivers.size,
  }
  // Prune older than 7 days
  const cutoff = new Date()
  cutoff.setDate(cutoff.getDate() - 7)
  const cutoffIso = cutoff.toISOString().slice(0, 10)
  const pruned = {}
  for (const [date, data] of Object.entries(daily)) {
    if (date >= cutoffIso) pruned[date] = data
  }
  state.daily_counts = pruned
  return state
}

function compute7DaySummary(state) {
  const daily = state.daily_counts || {}
  const driverDays = {}
  let totalViolationDays = 0

  for (const [date, dayData] of Object.entries(daily).sort()) {
    const drivers = dayData.drivers || []
    totalViolationDays += drivers.length
    for (const name of drivers) {
      driverDays[name] = (driverDays[name] || 0) + 1
    }
  }

  const sorted = Object.entries(driverDays)
    .sort(([, a], [, b]) => b - a)
    .reduce((acc, [name, days]) => { acc[name] = days; return acc }, {})

  return {
    total_violation_days: totalViolationDays,
    drivers: sorted,
    days_tracked: Object.keys(daily).length,
  }
}

// ── Violation detection ────────────────────────────────────────────────────────

function detectViolations(drivers, alerts) {
  const today = new Date().toISOString().slice(0, 10)

  // Non-compliant / expiring-soon drivers
  const nonCompliant = []
  for (const d of drivers) {
    if (!d.active) continue
    const cs = d.complianceStatus || 'UNKNOWN'
    if (cs === 'NON_COMPLIANT' || cs === 'EXPIRING_SOON') {
      nonCompliant.push({
        name: d.name || 'Unknown',
        driver_id: d.id,
        status: cs,
        driver_type: d.type || 'UNCLASSIFIED',
        cdl_expiration: d.cdlExpiration || null,
        med_card_expiration: d.medCardExpiration || null,
        assigned_truck_id: d.assignedTruckId || null,
        phone: d.phone || null,
      })
    }
  }

  // Unacknowledged severe alerts (driver only)
  const severeAlerts = []
  for (const a of alerts) {
    if (a.acknowledged) continue
    if (a.resolvedAt) continue
    if (a.entityType !== 'DRIVER') continue
    const sev = a.severity || ''
    if (sev !== 'CRITICAL' && sev !== 'EXPIRED') continue
    severeAlerts.push({
      driver_name: a.entityName || 'Unknown',
      driver_id: a.entityId,
      alert_id: a.id,
      document_type: a.documentType,
      document_title: a.documentTitle || null,
      expiration_date: a.expirationDate || null,
      severity: sev,
      created_at: a.createdAt || null,
    })
  }

  // Combined violation set
  const violationNames = new Set()
  const violationDetails = []

  for (const d of nonCompliant) {
    violationNames.add(d.name)
    violationDetails.push({
      driver: d.name,
      type: 'compliance_status',
      detail: `Status: ${d.status}`,
      driver_type: d.driver_type,
      cdl_exp: d.cdl_expiration,
      med_exp: d.med_card_expiration,
    })
  }

  const seenAlertKeys = new Set()
  for (const a of severeAlerts) {
    const name = a.driver_name
    if (!violationNames.has(name)) violationNames.add(name)
    const key = `${name}|${a.document_type}`
    if (!seenAlertKeys.has(key)) {
      seenAlertKeys.add(key)
      const expStr = a.expiration_date ? a.expiration_date.slice(0, 10) : 'N/A'
      violationDetails.push({
        driver: name,
        type: 'document_alert',
        detail: `${a.severity}: ${a.document_title || a.document_type} (expires ${expStr})`,
        alert_id: a.alert_id,
      })
    }
  }

  const activeCount = drivers.filter(d => d.active).length

  return {
    violation_drivers: [...violationNames].sort(),
    violation_count: violationNames.size,
    non_compliant: nonCompliant,
    severe_alerts: severeAlerts,
    violations: violationDetails,
    total_drivers: activeCount,
  }
}

// ── Slack message formatting ───────────────────────────────────────────────────

function colorForSeverity(severity) {
  if (severity === 'EXPIRED') return '🔴'
  if (severity === 'CRITICAL') return '🟠'
  if (severity === 'URGENT') return '🟡'
  return '⚪'
}

function emojiForTotal(total) {
  if (total === 0) return '✅'
  if (total <= 2) return '⚠️'
  return '🚨'
}

function format7DaySummary(summary) {
  const lines = [
    '*7-Day Rolling Violation Count:*',
    `  • Total violation-days: ${summary.total_violation_days}`,
    `  • Days tracked: ${summary.days_tracked}`,
  ]
  const drivers = Object.entries(summary.drivers)
  if (drivers.length > 0) {
    lines.push('  • Per-driver breakdown:')
    for (const [name, days] of drivers) {
      const bar = '█'.repeat(days)
      lines.push(`    ${name}: ${days}d ${bar}`)
    }
  } else {
    lines.push('  • No violations in the last 7 days. 🎉')
  }
  return lines.join('\n')
}

function formatSlackMessage(results, summary7d) {
  const today = new Date().toISOString().slice(0, 10)
  const vcount = results.violation_count
  const lines = [
    `${emojiForTotal(vcount)} *HOS / ELD Compliance Watchdog* — ${today}`,
  ]

  if (vcount === 0) {
    lines.push('✅ No HOS compliance violations detected in the last 24 hours.')
    lines.push(`_${results.total_drivers} active drivers monitored._`)
    lines.push('')
    lines.push(format7DaySummary(summary7d))
    return lines.join('\n')
  }

  lines.push(
    `*${vcount} driver(s)* with compliance issues detected ` +
    `(out of ${results.total_drivers} active).`
  )
  lines.push('')

  // Per-driver violation details
  for (const v of results.violations) {
    lines.push(`• *${v.driver}*: ${v.detail}`)
  }

  // Non-compliant drivers
  if (results.non_compliant.length) {
    lines.push('')
    lines.push('*Non-Compliant / Expiring Drivers:*')
    for (const d of results.non_compliant) {
      const extras = []
      if (d.cdl_expiration) extras.push(`CDL: ${d.cdl_expiration}`)
      if (d.med_card_expiration) extras.push(`Med: ${d.med_card_expiration}`)
      const extrasStr = extras.length ? ` ${extras.join(' | ')}` : ''
      lines.push(`  • ${d.name} — ${d.status} (${d.driver_type})${extrasStr}`)
    }
  }

  // Severe alerts
  if (results.severe_alerts.length) {
    lines.push('')
    lines.push('*Unacknowledged CRITICAL / EXPIRED Alerts (last 24h):*')
    const shown = results.severe_alerts.slice(0, 10)
    for (const a of shown) {
      const exp = a.expiration_date ? a.expiration_date.slice(0, 10) : '?'
      lines.push(
        `  ${colorForSeverity(a.severity)} ${a.driver_name}: ` +
        `${a.document_title || a.document_type} expires ${exp}`
      )
    }
    if (results.severe_alerts.length > 10) {
      lines.push(`  … +${results.severe_alerts.length - 10} more`)
    }
  }

  lines.push('')
  lines.push(format7DaySummary(summary7d))

  return lines.join('\n')
}

// ── CLI ────────────────────────────────────────────────────────────────────────

function parseArgs(argv) {
  const args = {}
  for (const a of argv) {
    if (a === '--json') args.json = true
    else if (a === '--slack') args.slack = true
    else if (a === '--quiet') args.quiet = true
  }
  return args
}

async function main() {
  const cli = parseArgs(process.argv.slice(2))

  const email = process.env.BCAT_EMAIL
  const password = process.env.BCAT_PASSWORD
  if (!email || !password) {
    console.error('Error: BCAT_EMAIL and BCAT_PASSWORD environment variables are required.')
    console.error('       Source them from ~/.hermes/bcat-ops-credentials.env')
    process.exit(1)
  }

  console.error(`Authenticating as ${email}...`)
  await signIn({ username: email, password })
  const session = await fetchAuthSession()
  const idToken = session.tokens?.idToken?.toString()
  if (!idToken) {
    console.error('No ID token after sign-in')
    process.exit(1)
  }

  console.error('Fetching drivers...')
  const driversResult = await callAppSync(LIST_DRIVERS_QUERY, {}, idToken)
  if (driversResult.errors) throw new Error(`GraphQL error: ${driversResult.errors[0].message}`)
  const drivers = driversResult.data.listDrivers.items || []
  console.error(`  Fetched ${drivers.length} driver records.`)

  console.error('Fetching compliance alerts...')
  const alertsResult = await callAppSync(LIST_ALERTS_QUERY, {}, idToken)
  if (alertsResult.errors) throw new Error(`GraphQL error: ${alertsResult.errors[0].message}`)
  const alerts = alertsResult.data.listComplianceAlerts.items || []
  console.error(`  Fetched ${alerts.length} alert records.`)

  const results = detectViolations(drivers, alerts)
  console.error(`  Detected ${results.violation_count} driver(s) with violations.`)

  // 7-day rolling state
  const state = loadState()
  const todayIso = new Date().toISOString().slice(0, 10)
  updateRollingCounts(state, todayIso, new Set(results.violation_drivers))
  saveState(state)

  const summary7d = compute7DaySummary(state)

  if (cli.json) {
    console.log(JSON.stringify({
      date: todayIso,
      violations: results,
      seven_day_rolling: summary7d,
    }, null, 2))
  } else if (cli.slack) {
    // Quiet mode: only output if violations exist
    if (cli.quiet && results.violation_count === 0) return
    console.log(formatSlackMessage(results, summary7d))
  } else {
    // Human-readable
    console.log('\nHOS / ELD Compliance Watchdog Report')
    console.log('=====================================')
    console.log(`Date: ${todayIso}`)
    console.log(`Active drivers: ${results.total_drivers}`)
    console.log(`Drivers in violation: ${results.violation_count}`)
    console.log()

    if (results.violation_count === 0) {
      console.log('✅ No compliance violations detected.')
    } else {
      for (const v of results.violations) {
        console.log(`  • ${v.driver}: ${v.detail}`)
      }
    }

    console.log()
    console.log(format7DaySummary(summary7d))
  }
}

main().catch(err => {
  console.error(err)
  process.exit(1)
})

File: scripts/detect_hos_violations.py
#!/usr/bin/env python3
"""HOS/ELD Compliance Watchdog — queries bcat-ops for driver compliance alerts,
tracks 7-day rolling violation counts, and posts violations to Slack.

Read-only. Queries the bcat-ops AppSync GraphQL API for:
  - Unacknowledged ComplianceAlerts (last 24h)
  - Drivers with NON_COMPLIANT or EXPIRING_SOON compliance status
Tracks daily violation counts per driver in a 7-day rolling state file.

Authentication: Cognito USER_PASSWORD_AUTH using BCAT_EMAIL / BCAT_PASSWORD env vars
(same auth as detect_fuel_anomalies.py).

Usage:
    BCAT_EMAIL=ryne@bcatcorp.com BCAT_PASSWORD=... python3 detect_hos_violations.py
    BCAT_EMAIL=... BCAT_PASSWORD=... python3 detect_hos_violations.py --slack
    BCAT_EMAIL=... BCAT_PASSWORD=... python3 detect_hos_violations.py --json
"""

import json
import os
import sys
import datetime
import argparse
from typing import Optional
from pathlib import Path

import boto3
from botocore.exceptions import ClientError

# ── Configuration ────────────────────────────────────────────────────────────────
APPSYNC_URL = "https://2ivs344vbvglblxbmjdd7pgi2q.appsync-api.us-east-1.amazonaws.com/graphql"
USER_POOL_ID = "us-east-1_IbPKPNJC9"
CLIENT_ID = "1qr651dshigrv8a6c338d4f709"
AWS_REGION = "us-east-1"

# State file for 7-day rolling violation tracking
STATE_FILE = Path.home() / ".hermes/state/hos-watchdog-state.json"

# ── GraphQL queries ──────────────────────────────────────────────────────────────

# Get all unacknowledged compliance alerts (the set is small)
LIST_ALERTS_QUERY = """
query ListComplianceAlerts {
  listComplianceAlerts(limit: 1000) {
    items {
      id entityType entityId entityName documentType documentTitle
      complianceDocumentId expirationDate severity acknowledged
      acknowledgedBy acknowledgedAt resolvedAt createdAt updatedAt
    }
  }
}
"""

# Get all drivers (with compliance status)
LIST_DRIVERS_QUERY = """
query ListDrivers {
  listDrivers(limit: 200) {
    items {
      id name active type driverType onboardingStatus complianceStatus
      cdl cdlExpiration medCardExpiration hireDate
      assignedTruckId email phone createdAt updatedAt
    }
  }
}
"""


def authenticate(email: str, password: str) -> str:
    """Sign in to Cognito and return the ID token."""
    client = boto3.client("cognito-idp", region_name=AWS_REGION)

    try:
        resp = client.initiate_auth(
            AuthFlow="USER_PASSWORD_AUTH",
            AuthParameters={"USERNAME": email, "PASSWORD": password},
            ClientId=CLIENT_ID,
        )
    except ClientError as e:
        code = e.response["Error"]["Code"]
        if code == "NotAuthorizedException":
            raise SystemExit(f"Authentication failed: invalid credentials ({e})") from e
        if code == "UserNotFoundException":
            raise SystemExit(f"Authentication failed: user not found ({e})") from e
        raise

    id_token = resp["AuthenticationResult"]["IdToken"]
    return id_token


def graphql_call(query: str, variables: dict, id_token: str) -> dict:
    """Make a GraphQL request to AppSync with Cognito auth."""
    import urllib.request

    body = json.dumps({"query": query, "variables": variables}).encode("utf-8")
    req = urllib.request.Request(
        APPSYNC_URL,
        data=body,
        headers={
            "Content-Type": "application/json",
            "Authorization": id_token,
        },
    )
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())


def fetch_drivers(id_token: str) -> list[dict]:
    """Fetch ALL drivers from AppSync."""
    result = graphql_call(LIST_DRIVERS_QUERY, {}, id_token)
    if result.get("errors"):
        raise SystemExit(f"GraphQL error: {result['errors'][0]['message']}")
    items = result.get("data", {}).get("listDrivers", {}).get("items", []) or []
    return items


def fetch_alerts(id_token: str) -> list[dict]:
    """Fetch ALL ComplianceAlert records from AppSync."""
    result = graphql_call(LIST_ALERTS_QUERY, {}, id_token)
    if result.get("errors"):
        raise SystemExit(f"GraphQL error: {result['errors'][0]['message']}")
    items = result.get("data", {}).get("listComplianceAlerts", {}).get("items", []) or []
    return items


# ── State file management ────────────────────────────────────────────────────────

def load_state() -> dict:
    """Load the 7-day rolling violation state from disk."""
    if STATE_FILE.exists():
        try:
            with open(STATE_FILE) as f:
                return json.load(f)
        except (json.JSONDecodeError, OSError):
            pass
    return {"daily_counts": {}, "last_updated": None}


def save_state(state: dict):
    """Persist the violation state to disk."""
    STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
    state["last_updated"] = datetime.datetime.now().isoformat()
    with open(STATE_FILE, "w") as f:
        json.dump(state, f, indent=2)


def update_rolling_counts(
    state: dict,
    today_iso: str,
    violation_drivers: set[str],
):
    """Add today's violations to the rolling state and prune days older than 7 days."""
    daily = state.get("daily_counts", {})

    # Add today's violations
    daily[today_iso] = {
        "drivers": sorted(violation_drivers),  # driver names
        "count": len(violation_drivers),
    }

    # Prune entries older than 7 days
    cutoff = (datetime.date.today() - datetime.timedelta(days=7)).isoformat()
    pruned = {k: v for k, v in daily.items() if k >= cutoff}
    state["daily_counts"] = pruned

    return state


def compute_7day_summary(state: dict) -> dict:
    """Compute per-driver 7-day rolling violation counts and total violation-days."""
    daily = state.get("daily_counts", {})
    driver_violation_days: dict[str, int] = {}
    total_violation_days = 0

    for date_str, day_data in sorted(daily.items()):
        drivers = day_data.get("drivers", [])
        total_violation_days += len(drivers)
        for driver_name in drivers:
            driver_violation_days[driver_name] = driver_violation_days.get(driver_name, 0) + 1

    return {
        "total_violation_days": total_violation_days,
        "drivers": dict(sorted(driver_violation_days.items(), key=lambda x: -x[1])),
        "days_tracked": len(daily),
    }


# ── Violation detection ──────────────────────────────────────────────────────────

def detect_violations(drivers: list[dict], alerts: list[dict]) -> dict:
    """Analyze drivers and alerts to find HOS/compliance violations.

    A driver is considered in violation if:
      - complianceStatus is NON_COMPLIANT or EXPIRING_SOON
      - Has unacknowledged CRITICAL or EXPIRED alerts
    """
    now = datetime.datetime.now(datetime.timezone.utc)
    today_iso = datetime.date.today().isoformat()
    yesterday_iso = (datetime.date.today() - datetime.timedelta(days=1)).isoformat()

    # ── Find non-compliant/expiring drivers ──
    non_compliant_drivers = []
    for d in drivers:
        if not d.get("active"):
            continue
        cs = d.get("complianceStatus", "UNKNOWN")
        if cs in ("NON_COMPLIANT", "EXPIRING_SOON"):
            non_compliant_drivers.append({
                "name": d.get("name", "Unknown"),
                "driver_id": d.get("id"),
                "status": cs,
                "driver_type": d.get("driverType") or "UNCLASSIFIED",
                "cdl_expiration": d.get("cdlExpiration"),
                "med_card_expiration": d.get("medCardExpiration"),
                "assigned_truck_id": d.get("assignedTruckId"),
                "phone": d.get("phone"),
            })

    # ── Find unacknowledged severe alerts ──
    severe_alerts = []
    for a in alerts:
        if a.get("acknowledged"):
            continue
        if a.get("resolvedAt"):
            continue
        severity = a.get("severity", "")
        if severity not in ("CRITICAL", "EXPIRED"):
            continue
        # Only driver alerts (not truck)
        if a.get("entityType") != "DRIVER":
            continue
        severe_alerts.append({
            "driver_name": a.get("entityName", "Unknown"),
            "driver_id": a.get("entityId"),
            "alert_id": a.get("id"),
            "document_type": a.get("documentType"),
            "document_title": a.get("documentTitle"),
            "expiration_date": a.get("expirationDate"),
            "severity": severity,
            "created_at": a.get("createdAt"),
        })

    # ── Combine into unified violation list ──
    # Build a set of driver names in violation
    violation_names: set[str] = set()
    violation_details = []

    for d in non_compliant_drivers:
        violation_names.add(d["name"])
        violation_details.append({
            "driver": d["name"],
            "type": "compliance_status",
            "detail": f"Status: {d['status']}",
            "driver_type": d["driver_type"],
            "cdl_exp": d.get("cdl_expiration"),
            "med_exp": d.get("med_card_expiration"),
        })

    seen_alert_drivers = set()
    for a in severe_alerts:
        name = a["driver_name"]
        if name not in violation_names:
            violation_names.add(name)
        if (name, a["document_type"]) not in seen_alert_drivers:
            seen_alert_drivers.add((name, a["document_type"]))
            exp_str = a.get("expiration_date", "")[:10] if a.get("expiration_date") else "N/A"
            violation_details.append({
                "driver": name,
                "type": "document_alert",
                "detail": f"{a['severity']}: {a['document_title'] or a['document_type']} (expires {exp_str})",
                "alert_id": a["alert_id"],
            })

    return {
        "violation_drivers": sorted(violation_names),
        "violation_count": len(violation_names),
        "non_compliant": non_compliant_drivers,
        "severe_alerts": severe_alerts,
        "violations": violation_details,
        "total_drivers": len([d for d in drivers if d.get("active")]),
    }


# ── Slack message formatting ─────────────────────────────────────────────────────

def color_for_severity(severity: str) -> str:
    if severity == "EXPIRED":
        return "🔴"
    if severity == "CRITICAL":
        return "🟠"
    if severity == "URGENT":
        return "🟡"
    return "⚪"


def emoji_for_total(total: int) -> str:
    if total == 0:
        return "✅"
    if total <= 2:
        return "⚠️"
    return "🚨"


def format_slack_message(results: dict, _7day_summary: dict) -> str:
    """Format violations as a Slack markdown message."""
    today = datetime.date.today().isoformat()
    violations = results["violations"]
    vcount = results["violation_count"]

    lines = [
        f"{emoji_for_total(vcount)} *HOS / ELD Compliance Watchdog* — {today}",
    ]

    if vcount == 0:
        lines.append("✅ No HOS compliance violations detected in the last 24 hours.")
        lines.append(f"_{results['total_drivers']} active drivers monitored._")
        # Still show 7-day summary
        lines.append("")
        lines.append(format_7day_summary_lines(_7day_summary))
        return "\n".join(lines)

    lines.append(
        f"*{vcount} driver(s)* with compliance issues detected "
        f"(out of {results['total_drivers']} active)."
    )
    lines.append("")

    # Per-driver violation details
    for v in violations:
        lines.append(
            f"• *{v['driver']}*: {v['detail']}"
        )

    # Non-compliant drivers section
    if results["non_compliant"]:
        lines.append("")
        lines.append("*Non-Compliant / Expiring Drivers:*")
        for d in results["non_compliant"]:
            extra = []
            if d.get("cdl_expiration"):
                extra.append(f"CDL: {d['cdl_expiration']}")
            if d.get("med_card_expiration"):
                extra.append(f"Med: {d['med_card_expiration']}")
            extras = " | ".join(extra) if extra else ""
            lines.append(f"  • {d['name']} — {d['status']} ({d['driver_type']}) {extras}")

    # Unacknowledged severe alerts
    if results["severe_alerts"]:
        lines.append("")
        lines.append("*Unacknowledged CRITICAL / EXPIRED Alerts:*")
        for a in results["severe_alerts"][:10]:  # cap at 10
            exp = a.get("expiration_date", "")[:10] if a.get("expiration_date") else "?"
            lines.append(
                f"  {color_for_severity(a['severity'])} {a['driver_name']}: "
                f"{a['document_title'] or a['document_type']} expires {exp}"
            )
        if len(results["severe_alerts"]) > 10:
            lines.append(f"  … +{len(results['severe_alerts']) - 10} more")

    # 7-day rolling summary
    lines.append("")
    lines.append(format_7day_summary_lines(_7day_summary))

    return "\n".join(lines)


def format_7day_summary_lines(summary: dict) -> str:
    """Format the 7-day rolling summary section."""
    lines = [
        "*7-Day Rolling Violation Count:*",
        f"  • Total violation-days: {summary['total_violation_days']}",
        f"  • Days tracked: {summary['days_tracked']}",
    ]

    if summary["drivers"]:
        lines.append("  • Per-driver breakdown:")
        for driver, days in summary["drivers"].items():
            bar = "█" * days
            lines.append(f"    {driver}: {days}d {bar}")
    else:
        lines.append("  • No violations in the last 7 days. 🎉")

    return "\n".join(lines)


def format_human_output(results: dict, _7day_summary: dict) -> str:
    """Human-readable text output."""
    vcount = results["violation_count"]
    lines = [
        "HOS / ELD Compliance Watchdog Report",
        "=====================================",
        f"Date: {datetime.date.today().isoformat()}",
        f"Active drivers: {results['total_drivers']}",
        f"Drivers in violation: {vcount}",
        "",
    ]

    if vcount == 0:
        lines.append("✅ No compliance violations detected.")
    else:
        for v in results["violations"]:
            lines.append(f"  • {v['driver']}: {v['detail']}")

    lines.append("")
    lines.append("7-Day Rolling Summary:")
    lines.append(f"  Total violation-days: {_7day_summary['total_violation_days']}")
    lines.append(f"  Days tracked: {_7day_summary['days_tracked']}")
    if _7day_summary["drivers"]:
        for driver, days in _7day_summary["drivers"].items():
            lines.append(f"    {driver}: {days} day(s)")
    else:
        lines.append("  No violations in last 7 days.")

    return "\n".join(lines)


# ── Main ─────────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="HOS/ELD compliance watchdog for bcat-ops")
    parser.add_argument("--json", action="store_true", help="Output raw JSON")
    parser.add_argument("--slack", action="store_true", help="Output Slack-formatted markdown")
    parser.add_argument("--quiet", action="store_true", help="Silent on no violations (for cron)")
    args = parser.parse_args()

    email = os.environ.get("BCAT_EMAIL")
    password = os.environ.get("BCAT_PASSWORD")
    if not email or not password:
        print("Error: BCAT_EMAIL and BCAT_PASSWORD environment variables are required.", file=sys.stderr)
        print("       Source them from ~/.hermes/bcat-ops-credentials.env", file=sys.stderr)
        sys.exit(1)

    print(f"Authenticating as {email}...", file=sys.stderr)
    id_token = authenticate(email, password)

    print("Fetching drivers...", file=sys.stderr)
    drivers = fetch_drivers(id_token)
    print(f"  Fetched {len(drivers)} driver records.", file=sys.stderr)

    print("Fetching compliance alerts...", file=sys.stderr)
    alerts = fetch_alerts(id_token)
    print(f"  Fetched {len(alerts)} alert records.", file=sys.stderr)

    # Detect violations
    results = detect_violations(drivers, alerts)
    print(f"  Detected {results['violation_count']} driver(s) with violations.", file=sys.stderr)

    # Load state, update rolling counts, save
    state = load_state()
    today_iso = datetime.date.today().isoformat()
    state = update_rolling_counts(state, today_iso, set(results["violation_drivers"]))
    save_state(state)

    # Compute 7-day summary
    _7day_summary = compute_7day_summary(state)

    # Output
    if args.json:
        output = {
            "date": today_iso,
            "violations": results,
            "seven_day_rolling": _7day_summary,
        }
        print(json.dumps(output, indent=2, default=str))
    elif args.slack:
        msg = format_slack_message(results, _7day_summary)
        print(msg)
    else:
        # Quiet mode: only output if violations exist (for cron deliveries)
        if args.quiet and results["violation_count"] == 0:
            return
        print(format_human_output(results, _7day_summary))


if __name__ == "__main__":
    main()
